### PR TITLE
feat: supporting $ref in create new config

### PR DESCRIPTION
**Pull request by almog8k**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 1]
```diff
@@ -55,30 +64,58 @@ export const Step2AddConfig: React.FC<Step2AddConfigProps> = ({ onDataChange, on
...
+    } catch (error) {
+      if (error instanceof ApiError) {
+        setNetworkError(error.message);
+      } else {
+        setNetworkError(undefined);
+      }
+    } finally {
+      setIsLoading(false);
+    }
+
```
Comment (CptSchnitz) [inline src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:104]: if there is a network error, fail the validation

File: src/utils/monaco/refHandler.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  const configResponse =
+    cacheConfig ??
+    (await queryClient.fetchQuery({
+      queryKey: [getVersionedConfig.name, config.name, config.version],
+      queryFn: () => getVersionedConfig(config),
+      queryHash: queryHash,
+    }));
+
+  return configResponse;
+}
```
Comment (CptSchnitz) [inline src/utils/monaco/refHandler.ts:93]: move to relevant file

**Review by CptSchnitz — CHANGES_REQUESTED** [review]
- the bar at the side for showing errors is too narrow on my screen.
- config not found should be validation error

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 3]
```diff
@@ -17,35 +20,41 @@ type Step2AddConfigProps = {
...
 export const Step2AddConfig: React.FC<Step2AddConfigProps> = ({ onDataChange, onJsonStringChange, schemaId, initialJsonStringData }) => {
-  // const fetchSchema = useCallback(() => getSchema({ id: schemaId, shouldDereference: false }), [schemaId]);
-  const fetchSchemaDereference = useCallback(() => getSchema({ id: schemaId, shouldDereference: true }), [schemaId]);
-
-  // const schemaWithRefsRes = useQuery({
-  //   queryKey: ['getSchema'],
-  //   queryFn: fetchSchema,
-  //   enabled: !!schemaId,
-  // });
+  const monaco = useMonaco();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:24]: done

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 4]
```diff
@@ -17,35 +20,41 @@ type Step2AddConfigProps = {
...
 
+  const fetchSchemaDereference = useCallback(() => getSchema({ id: schemaId, shouldDereference: true }), [schemaId]);
   const { data: schema, isSuccess } = useQuery({
     queryKey: ['getSchemaWitheRefs'],
     queryFn: fetchSchemaDereference,
     enabled: !!schemaId,
   });
 
-  const monaco = useMonaco();
+  const [isLoading, setIsLoading] = useState<boolean>(false);
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:33]: done

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
+import { useCallback, useEffect, useMemo, useState } from 'react';
 import { useMonaco } from '@monaco-editor/react';
 import { ConfigData } from '../createConfig.types';
-import { registerRefSnippet } from '../../../utils/monaco/snippetsRegistration/registerRefSnippet';
+import { registerRefSnippet } from '../../../utils/monaco/register/registerRefSnippet';
+import { dereference, isRef } from '../../../utils/monaco/refHandler';
+import { ErrorCard } from '../../../components/errorCard/errorCard';
+import { registerRefHoverProvider } from '../../../utils/monaco/register/registerRefHoverProvider';
 
 const ajvInstance = new ajv();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:15]: done

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 6]
```diff
@@ -17,35 +20,41 @@ type Step2AddConfigProps = {
...
+  const fetchSchemaDereference = useCallback(() => getSchema({ id: schemaId, shouldDereference: true }), [schemaId]);
   const { data: schema, isSuccess } = useQuery({
     queryKey: ['getSchemaWitheRefs'],
     queryFn: fetchSchemaDereference,
     enabled: !!schemaId,
   });
 
-  const monaco = useMonaco();
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [initProgressBar, setInitProgressBar] = useState<boolean>(false);
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:34]: moved inside MonacoEditor

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 7]
```diff
@@ -55,30 +64,58 @@ export const Step2AddConfig: React.FC<Step2AddConfigProps> = ({ onDataChange, on
...
+    onJsonStringChange(value);
+
     try {
-      onJsonStringChange(value);
-      const json = JSON.parse(value ?? '');
-      const isValid = await validate(json);
+      if (value !== undefined && isRef(value)) {
+        setIsLoading(true);
 
+        jsonDereferenced = await dereference(value);
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:92]: named changed

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 8]
```diff
@@ -55,30 +64,58 @@ export const Step2AddConfig: React.FC<Step2AddConfigProps> = ({ onDataChange, on
...
+    const snippetProvider = registerRefSnippet(monaco);
+    const hoverProvider = registerRefHoverProvider(monaco);
+
+    return () => {
+      snippetProvider.dispose();
+      hoverProvider.dispose();
+    };
   }, [diagnosticOptions, monaco]);
 
   const handleChange = async (value: string | undefined) => {
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:84]: changed

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx (hunk 1 above)
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:104]: done

File: src/utils/monaco/refHandler.ts (hunk 2 above)
Comment (almog8k) [reply src/utils/monaco/refHandler.ts]: done

File: src/api/services/configDataFetcher.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    (await queryClient.fetchQuery({
+      queryKey: [getVersionedConfig.name, config.name, config.version],
+      queryFn: () => getVersionedConfig(config),
+      queryHash: queryHash,
+    }));
+
+  return configResponse;
+}
+
+export function createGetVersionedConfigQueryHash(config: GetVersionedConfigData): string {
```
Comment (CptSchnitz) [inline src/api/services/configDataFetcher.ts:18]:
```suggestion
export function getVersionedConfigQueryHash(config: GetVersionedConfigData): string {
```

File: src/api/services/configDataFetcher.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  return `${getVersionedConfig.name}${config.name}${config.version}`;
```
Comment (CptSchnitz) [inline src/api/services/configDataFetcher.ts:19]: add separator between the parameters

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

**Review by CptSchnitz — APPROVED** [review]

**Pull request merged**